### PR TITLE
cse prototype

### PR DESCRIPTION
**Pull request by yushangdi**
CSE algorithm is in cse.py and unit tests are in unittest_cse.py

The CSE algorithm:

The logic is basically just hashing each node’s target, args, and kwargs to find opportunities for elimination. 

It returns a new graph with CSE applied to the input graph.

A mapping from node in the old graph to node in the new graph is stored in env. 

Before hashing, each element in args and kwargs are checked and substituted by their mapping in env if available.

A node is substitued if there exists a node that has the same hash and the same args and kwargs (to avoid problems caused by hash collision). One caveat here is that if hash collision happens, only one set of nodes are eliminated. e.g. if hash(node1)=hash(node2) = hash(node3)=hash(node4), but node1=node2 != node3=node4, node 2 will be eliminated, but node 4 will not. 

1. when substituting, we check (nested-)list arguments and kwargs to substitute nodes in last level of lists
2. has support for checking kwargs
3. when checking if two nodes are the same, check (nested-)list arguments and kwargs to compare nodes in last level of lists
4. Convert lists and dictionaries into tuples when hashing. Dictionaries are sorted by keys before conversion. Now we only use the python default hash function

Unit tests are added in unittest_cse.py

The tests test for:

1. the number of nodes decrease by an expected number
2. a second CSE does not change the number of nodes anymore
3. the results of computations are the same

Currently the test has:

1. a simple test of not changing the graph when there is no available reduction
2. should not CSE way random generators ([aten.rand_like, aten.rand, aten.randint, aten.randn])
3. works for empty function
4. no args
5. a single arg
6. multiple args
7. args has nested lists
8. args where argument is the default value
9. args where argument is not the default value
10. a single duplicated node
11. multiple duplicated node
12. no kwargs
13. has kwargs

### CODE REVIEW

**Comment by yushangdi** [comment]
Changed the CSE substitution to use pytree flatten.

File: cse/cse.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            arg_list[i] = env[v]
+    return tuple(arg_list), spec
+
+# return a new graph with CSE applied to the input graph
+# env stores a mapping from node in the old graph to node in the new graph
+# The placehold, output, and get_attr nodes are copied to the new grpah without change
+# The call nodes (call_function) are hashed to check if they
+# have an equivalent node in the graph. If so, this node will not be copied, and a mapping
+# to the duplicated node is stored in env
+def modify(fx_g: torch.fx.graph.Graph):
```
[the remark this answers is not in the document]
Comment (yushangdi) [reply cse/cse.py]: renamed to fx_graph_cse

File: cse/unittest_cse.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    def test_kwarg(self):
+        def f(x):
+            a = torch.ones_like(x)
+            b = torch.ones_like(x)
+            return a + b
+        t = torch.randn(2, 2)
+        check(f, t, 1)
+
+
+class RandomOpTestCase(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (yushangdi) [reply cse/unittest_cse.py]: now changed to run 30 times

File: cse/unittest_cse.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
+import torch
```
[the remark this answers is not in the document]
Comment (yushangdi) [reply cse/unittest_cse.py]: moved

File: cse/cse.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
+import torch
```
[the remark this answers is not in the document]
Comment (yushangdi) [reply cse/cse.py]: moved

**Comment by yushangdi** [comment]
Benchmarks are added to measure the performance improvement.

The benchmarks are setup such that for each function `f`, we benchmark fx_g and new_g **without passing through torchscript**. The reason is that torchscript also has a CSE optimization. 

```python
import torch.fx as fx
from functorch import make_fx
from functorch._src.cse import fx_graph_cse

    fx_g = make_fx(f)(inp)

    new_g = fx_graph_cse(fx_g.graph)
    new_g = fx.GraphModule(fx_g, new_g)
```

For each benchmark, we run the function 5 times to warmup, and then measure another 5 runs of the function. We record the average time of the 5 runs.

**Review by Chillee — APPROVED** [review]
LGTM!

File: cse/benchmark.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
+import torch
```
Comment (Chillee) [inline cse/benchmark.py:1]: Should probably remove this top-line folder. You can move it into `benchmarks` if you like.

Comment (yushangdi) [reply cse/benchmark.py]: moved

File: benchmarks/chrome_trace_parser.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
+#!/usr/bin/env python
```
Comment (Chillee) [inline benchmarks/chrome_trace_parser.py:1]: btw, let's include this in a follow up PR.
